Support More Than one Token - v1

m4rj4nzadeh opened the pull request:
- Change req body type to array ( change client side ).
- Separate Func & Ctrl in harvest for supporting an array.
- Promise all for processing the request body for each token type.
- Basic stage for testing v2 revision of harvest API.

m4rj4nzadeh committed:
Support More Than one Token

- Change req body type to array ( change client side )
- Separate Func & Ctrl in harvest for supporting array
- Promise all for processing the request body for each

m4rj4nzadeh committed:
Changes on submitJob & PendingRewards

- Adding /v2 routes for submitJob & Pending Rewards
- Clone and add related parts to new components

BenjaminGrabow commented on components/harvestTx.ts:
```diff
@@ -15,30 +15,38 @@ if (!BLOCKFROST_URL || !BLOCKFROST_KEY || !NETWORK) {
-  let address = parsedBody.address
-  console.log(`Body: ${body}`)
+  console.log(parsedBody)
+  const harvestTxRes = await Promise.all(parsedBody.map(async (harvestReq) => {
+    return await harvestTX(prisma, harvestReq);
+  }));
```
We should send all 3 tokens in 1 transaction and not 3 transactions so that the user in the frontend only needs to sign 1 message instead of 3 messages (better UX).

BenjaminGrabow commented on components/pendingRewards.ts:
```diff
@@ -0,0 +1,96 @@
+}
+
+const pendingRewardsCtrl = async (prisma: PrismaClient, body: string) => {
+    const parsedBody: RewardReqBody = JSON.parse(body)
+    let poolI = parsedBody.poolIndex
+    let address = parsedBody.address
+    const poolargs = stakingpools[poolI]
+
+    let pool = await prisma.pool.findFirst({
+        where: {
+            poolIndex: poolI
+        }
+    })
+    if (!pool) {
+        pool = await prisma.pool.create({
+            data: {
+                poolIndex: poolI
```
Did you change any code in the pendingRewards or submit controllers, or did you just relocate the controllers into different files?

BenjaminGrabow replied:
Ah I see, you created them new as v2. Are there any differences between the old controllers? If yes very hard to recognize any changes compared to the old controllers (that was the reason for my question)

BenjaminGrabow commented on components/pendingRewards.ts:
```diff
@@ -0,0 +1,96 @@
+        let epochRatio = delta/432000000
+        let updatePay = Math.floor(epochRatio*(poolargs.poolInfo.rewardPerEpochQt))
+        let valueAdded: Assets = {}
+        valueAdded[poolargs.poolInfo.harvestUnit] = BigInt(updatePay)
+
+        await dbFromPool(poolI, poolargs, valueAdded)
+    } else {
+        console.log("We are not updating, just paying.")
+        console.log(Date.now() - lastUpdate.valueOf())
+    }
```
If we're going to start changing the controllers, I would love to see that we're handling promises more appropriately. Tons of promises are not getting caught. Don't see it as a task for now, the functionality part is more important, just a heads-up

m4rj4nzadeh committed:
Adding Quantity

- Adding quantity to harvest and pendingReward /v2
- multiple quantities to rewardPerEpochQt for calculating 3 token as a reward.

BenjaminGrabow commented on components/pendingRewards.ts:
```diff
@@ -53,7 +62,7 @@ const pendingRewardsCtrl = async (prisma: PrismaClient, body: string) => {
 
         let epochRatio = delta/432000000
-        let updatePay = Math.floor(epochRatio*(poolargs.poolInfo.rewardPerEpochQt))
+        let updatePay = Math.floor(epochRatio * ( quantity * (poolargs.poolInfo.rewardPerEpochQt)))
```
If the pool has more than 1 token we want to have the ability that each token has its own rewardPerEpochQt. (Example: token 1 has 500 rewardPerEpochQt, token 2 has 300 rewardPerEpochQt etc. ...

BenjaminGrabow replied:
@marjankhz I think it will be very helpful to test the full application. Then you can also verify if your solution works.

m4rj4nzadeh replied:
Thanks for the information. I already changed the way of progress to handle multiple tokens with different rewards and I will also have another push including some changes plus testing the flow.

m4rj4nzadeh committed:
Update harvestTx.ts

- Add each token processing for handling tokens.
